### PR TITLE
feat: gather v1

### PR DESCRIPTION
**Pull request by edisontim**
Introduces the base of a working loop with memory for iteration on GPT prompt and tweaking of relevance parameters

### CODE REVIEW

File: overlore/llm/constants.py [hunk 1]
```diff
@@ -1,88 +1,70 @@
...
-     Do note that all decisions are taken exclusively by the Lord of the realm.
-     As such although the characters can suggest future actions to be taken in an indirect manner, they cannot make the decision themselves.
+Use the greimas actantial model to put the characters in a role in a soap opera like plot.
+The only is the dialog between the characters. Format the conversation like so: <name>:<phrase>. You will end the conversation with: "!end of discussion!".
+Output a summary of the conversation, be specific about what happened and don't be vague.
+The characters will be given to you as input. You cannot create new characters on your own.
+All important decisions are taken exclusively by the Lord of the realm.
+As such although the characters can suggest future actions to be taken in an indirect manner, they cannot make the decision themselves.
 
-     The settings are as follows:
```
Comment (jobez) [inline overlore/llm/constants.py:53]: is there a rationale somewhere for this diff in the world system prompt?

File: overlore/llm/open_ai.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    ROLE,
+    SEX,
+    WORLD_SYSTEM_TEMPLATE,
+)
+from overlore.sqlite.constants import EventType
+from overlore.sqlite.types import StoredEvent
+from overlore.utils import get_enum_name_by_value, str_to_json
+
+
+class OpenAIHandler:
```
Comment (jobez) [inline overlore/llm/open_ai.py:25]: if agreed on we can create an issue for the backlog:

the 'nl' stuff can be refactored so we have a class that just handles the text formatting logic that can be used across different LLM services

File: overlore/llm/open_ai.py [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+    async def request_embedding(self, str_input: str) -> list[float]:
+        response = self.client.embeddings.create(
+            model="text-embedding-ada-002", input=str_input, encoding_format="float"
+        )
+        return response.data[0].embedding
+
+    async def generate_townhall_discussion(
+        self, realm_id: int, townhall_summaries: list[str], npc_list: list[Villager], events: list[StoredEvent]
+    ) -> str:
+        realms = Realms.instance()
```
Comment (jobez) [inline overlore/llm/open_ai.py:152]: can we initialize the class as having a realms instance, as opposed to constructing one per townhall gen req?

**Review by jobez — COMMENTED** [review]
🎁

File: overlore/llm/constants.py (hunk 1 above)
Comment (edisontim) [reply overlore/llm/constants.py:53]: I didn't really pay attention to what I was taking out/adding to the prompt at this point, was iterating and seeing where I would get the best results

File: overlore/llm/open_ai.py (hunk 2 above)
Comment (edisontim) [reply overlore/llm/open_ai.py:25]: Yeah completely agree on this

File: overlore/llm/open_ai.py (hunk 3 above)
Comment (edisontim) [reply overlore/llm/open_ai.py:152]: hmm can assign the instance to the class but the Realms class is a singleton so it's simply getting the object, maybe creating a copy of the same instance, but it's not creating a new one

Comment (jobez) [reply overlore/llm/open_ai.py:152]: re: singleton

having the singleton passed by init might make it easier to test for mocking

File: overlore/llm/open_ai.py (hunk 2 above)
Comment (jobez) [reply overlore/llm/open_ai.py:25]: issue'ed:https://github.com/The-Overlore/lore-machine/issues/35

File: overlore/llm/open_ai.py (hunk 3 above)
Comment (edisontim) [reply overlore/llm/open_ai.py:152]: Yeah true will change

**Review by cwastche — APPROVED** [review]
lgtm. Would probably be better to take out of the openAIHandler the nl methods.
Will create an issue for it, as it doesn't seem like a priority

**Pull request merged**